### ISSUE TITLE
Better Techs Proposal

### ISSUE THREAD

**Issue by gkatsev** [videojs/video.js#3734]
Having worked with Techs and Source Handlers for a while now, we recently found that there was a missing use case to our current support matrix.
The two main issues we were facing
1. we wanted to have a source handler or a tech that could use already preset source handlers or techs. 
For example, if we had a CMS that can put a video id in as a source for the video and we wanted a way to have a source handler/tech translate this video id into an HLS url and have it play back through the contrib-hls source handler.
2. we want to allow techs and source handlers to have a separate timeline that is displayed to the user.
This can be useful if you're working with a live playlist and want to allow the user to seek within it or you have server-side ads.

An interesting solution we were investigating was making source handlers behave like express middleware. During source selection, videojs would cycle through each of the source handlers giving each the source object and allow the source handler to modify it before it is handed to the next source handler. While this sounds like a good idea it only solves issue 1. To solve issue 2, we would need to modify the code a lot so that each method dealing with time was passed through each source handler in each direction. Plus, we already had a way to sit between what the underlying playback engine does and what the control bar and player produce: the Tech.

My proposal is that we update source handlers so that when you inherit from techs, you get to use its source handlers. If a source handler is registered on your tech and it applied to the same source as your parent tech, it would get chosen first. In addition, `"current"` timing methods will be added to the player and tech that the control bar will use -- I have them called `currentCurrentTime` and `currentDuration` in a local proof of concept, but we probably should come up with better names. These will default to use the normal `currentTime` and `duration` methods but if a tech wanted to provide an alternate timeline, they could override these methods. This is not to say that source handlers as middleware shouldn't happen but this proposal is to get the most benefit with the least amount of work.

In addition, work should be done to make techs that load asynchronously not break the player completely. Potentially, re-trying auto-setup on players with errors each time that `registerTech` is called.

In short:
* Allow source handlers to be inherited by Techs
* New Source Handlers can be registered to child techs and they will be chosen first over parent tech's source handlers
* Add new "current" timing methods like `currentCurrentTime` and `currentDuration` that are used by the control bar.
  * Default to the standard `currentTime` and `duration` methods.
  * Can be overridden by child techs to provide an alternate timeline for display
* Make asynchronously loaded techs work with auto-setup.

**Comment by gkatsev** [comment]
Would love to have some input from @bcvio @imbcmdth @heff @incompl @dmlap @misteroneill and whoever else wants to chime in.

**Comment by gkatsev** [comment]
A thing to add, after talking with @imbcmdth and @dmlap offline, I'm thinking that maybe the middleware idea won't be that hard to make. One thing that I was initially trying to do is keep/maintain backwards compatibility and have the opt-in be as part of source handlers. Now, we're thinking that maybe the middleware needs to be a separate thing from sourcehandlers with explicit opt-in again.
Not having to maintain backwards compatibility makes things easier, plus, if you strictly limit what powers are given to this middleware (with the ability to give it more functionality later on), we might be able to produce a working version in not a whole lot of time.
I still think the change I proposed above make sense, at least the source handler inheritance in techs, so, my aim would be to try and do both (source handler inheritance is pretty much done locally anyway 👅). So, in the *spirit of no compromises*, I propose that we do the new separate middleware and the source handler inheritance.

**Comment by heff** [comment]
> Add new "current" timing methods like currentCurrentTime and currentDuration that are used by the control bar.

It sounds like we'd be creating new special time APIs on the player that the control bar uses because we know about them. And someone building their own time display outside of the player might either get different time values than the control bar (because they're using the regular currentTime API), or would have to know about these special APIs. Am I understanding that right?

Otherwise this all sounds pretty good & interesting. I don't fully understand how the middleware would work yet. Would it be tech-specific or agnostic?

**Comment by gkatsev** [comment]
Yeah, I guess the idea with the new API methods is to make it so that you wouldn't need to write your own controls and we would make sure to document the new methods. Though, with the middleware idea, we wouldn't need this wouldn't be necessary.

The middleware idea will be a bit similar to how redux's middleware work. Each middleware will get a source object and a special version of the tech that allows the middleware to modify a certain subset of functionality, like `currentTime`, `setCurrentTime`, and `duration`. Then, the tech would chain together the setters from the first to the last and the getters from the last to the first.
So, for example, we would have the main Tech -> new timeline middleware -> HLS middleware -> video element. And so, setters would flow through the chain Tech->new timeline->HLS->video element, and getters would flow in the opposite direction: Tech<-new timeline<-HLS<-video.
And we can probably expand on it in the future by allowing the middleware to intercept events as well which would allow contrib-ads to turn into a middleware that can modify the timeline for ad controls and prevent events from getting through to the player and causing content flash.

**Comment by heff** [comment]
Nice, that seems like a cool approach.

Definitely would prefer to avoid special player API functions for this if it's at all avoidable.

**Comment by boushley** [comment]
I agree with @heff, having the special time APIs seems hacky and doesn't scale well.

@gkatsev the middleware idea sounds great. I'm interested in what the syntax might look like though. We'll have to be sure a middleware can intercept the things it cares about but be able to not worry about getters/setters/events it doesn't care about. This will make them easier to write and let us add new capabilities while keeping backwards comparability.

**Comment by gkatsev** [comment]
Yeah, I would prefer not to add a new API that is deprecated pretty much instantly. But my initial proposal was pragmatic one based on time to fix the usecase and the fact that we need this new usecase yesterday :P.

As for how the middleware will work, we're still working that out. And @boushley, yup, that's the plan. Basically make it so that `currentTime` and `duration` related stuff are available now but that other things can be added later on.
also, this new middleware idea is likely to be separate from source handlers completely but hopefully significantly more powerful.

**Comment by gkatsev** [comment]
middleware is in!
